Tasks/AzureRmWebAppDeployment/KuduUtility.ts

```diff
@@ -0,0 +1,89 @@
+import Q = require('q');
+import tl = require('vsts-task-lib/task');
+import path = require("path");
+import fs = require("fs");
+var request = require ('request');
+var archiver = require('archiver');
+
+
+/**
+ *  Deploys a zip based webapp package.
+ * 
+ *  @param  webAppPackage                  Zip file or folder for deployment
+ *  @param  virtualApplicationMappings     Mapping to get physical path for deployment
+ *  @param  publishingProfile              publish profile provides destination details for deployment
+ *  @param  virtualApplication             (Optional) Virtual application name 
+ */
+export async function deployWebAppPackage(webAppPackage: string, virtualApplicationMappings, publishingProfile,
+                             virtualApplication: string) {
+
+    var deferred = Q.defer<any>();	
+    // construct URL depending on virtualApplication or root of webapplication 
+    var physicalDeploymentPath = "/site/wwwroot";
+    var virtualPath = "/";
+    if (virtualApplication) {
+		virtualPath = "/" + virtualApplication;
+    }
+
+    for( var index in virtualApplicationMappings ){
+        var mapping = virtualApplicationMappings[index];
+        if( mapping.virtualPath == virtualPath){
+            physicalDeploymentPath = mapping.physicalPath;
+            break;
+        }
+    }
+	
+    tl.debug(tl.loc("Deployingwebapplicationatvirtualpathandphysicalpath", webAppPackage, virtualPath, physicalDeploymentPath));
+
+    var kuduDeploymentURL = "https://" + publishingProfile.publishUrl + "/api/zip/" + physicalDeploymentPath;
+    var basicAuthToken = 'Basic ' + new Buffer(publishingProfile.userName + ':' + publishingProfile.userPWD).toString('base64');
+    await fs.createReadStream(webAppPackage).pipe(request.put({ url: kuduDeploymentURL, headers: {
```

Comment:
Any way to wrap request.put into a promise so you don't need the error callback with the all the defer/reject code?  Is the webapppackage that big that it's important you pipe the stream?